test/expect-tests/dune_rpc/dune_rpc_tests.ml

```diff
@@ -233,3 +248,80 @@ let%expect_test "versioning public methods" =
     client: 20
     client: error invalid version
     server: finished. |}]
+
+let%test_module "subscription" =
+  (module struct
+    let sub_decl = { Dune_rpc_private.Sub.diff = Conv.string; name = "pulse" }
+
+    let version = (3, 0)
+
+    let init = init ~version ()
+
+    let rpc () = Handler.create ~on_init ~version ()
+
+    let server f =
+      let rpc = rpc () in
+      let () =
+        let on_subscribe _ = Fiber.return 42 in
+        let subscription _ _ sub = f sub in
+        Handler.subscription rpc sub_decl ~on_subscribe ~subscription
+      in
+      rpc
+
+    let run_client client on_next =
+      printfn "client: making subscription";
+      let* sub, stream = Client.subscribe client sub_decl in
+      let+ () =
+        Fiber.Stream.In.sequential_iter stream ~f:(fun a -> on_next a sub)
+      in
+      printfn "subscription terminated"
+
+    let%expect_test "subscription - client cancel" =
+      let client client =
+        let count = ref 0 in
+        run_client client (fun v active_sub ->
+            if !count = 2 then
+              Client.Subscription.cancel active_sub
+            else (
+              printfn "client: update %s received" v;
+              incr count;
+              Fiber.return ()
+            ))
+      in
+      let handler =
+        server (fun sub ->
+            let* () = Subscription.update sub "first" in
+            let* () = Subscription.update sub "second" in
+            let* () = Subscription.update sub "third" in
+            Subscription.finished sub)
+      in
+      test ~init ~client ~handler ();
+      [%expect
+        {|
+    client: making subscription
+    client: update first received
+    client: update second received
+    subscription terminated
+    server: finished. |}]
+
+    let%expect_test "subscription - server cancel" =
+      let client client =
+        run_client client (fun v _ ->
+            printfn "client: update %s received" v;
+            Fiber.return ())
+      in
+      let handler =
+        server (fun sub ->
+            let* () = Subscription.update sub "first" in
+            let* () = Subscription.update sub "second" in
+            Subscription.finish sub)
+      in
+      test ~init ~client ~handler ();
+      [%expect
+        {|
+    client: making subscription
+    client: update first received
+    client: update second received
+    subscription terminated
+    server: finished. |}]
```

Comment:
What ensures the ordering here? At the surface it looks like there's a race between the client finishing and the server finishing.

Reply:
> What ensures the ordering here?

The absence of IO makes this test fully deterministic (I hope).